File: jpp/4/2/DH/src/main/java/org/example/DHSetup.java
package org.example;

import java.util.Random;

public class DHSetup<T extends GF> {

  private final T generator;

  private T generateGenerator() {
    T dum = null;
    Random random = new Random();
    long number = random.nextLong(dum.getCharacteristic() - 1) + 1;
    return (T) new T(number);
  }

  public DHSetup() {
    this.generator = generateGenerator();
  }

  public T getGenerator() {
    return generator;
  }

  T power(T a, long b) {
    T temp = new T();
    if (b == 0) {
        return (T) new Object
    }
  }


}

File: jpp/4/3/final_ostatnie_do_oddania/ConsoleApp1/ConsoleApp1/DHSetup.cs

using System;

public class DHSetup<T> where T: GF, new()
{
   private T generator;
   
   private T generateGenerator(T dum)
   {
      Random random = new Random();
      int max = (int)dum.GetCharacteristic();
      int number = random.Next(1, max - 1);
      return (T) new GF (number);
   }

   public DHSetup(T dum)
   {
      this.generator = generateGenerator(dum);
   }

   public T getGenerator()
   {
      return generator;
   }
   
   public T power(T a, ulong b) {
      if (b == 0) {
         return (T) new GF (1);
      }
      T temp = power(a, b/2);
      if (b % 2 == 0) {
         return (T) (temp * temp);  
      } else {
         return (T) (a*temp*temp);
      }
    
   }
}
